fix(gnovm): Prevent use of blank identifier as Value or Type

notJoon opened the pull request:
# Description

Closes #1946 

The `isNamedConversion` function now includes a safety check to prevent the use of blank identifiers ("_") as values or types. If both `xt` and `t` are nil, the function assumes that a blank identifier is being used inappropriately and panics with an error message that includes the location of the issue.

## Variable Explanations

- `xt` (Expression Type): Represents the type of the right-hand side of an assignment or expression. It's the type resulting from evaluating an expression.
- `t` (Target Type): Represents the type of the left-hand side of an assignment. It's the variable or field that will receive the value.

Checks if a named conversion is needed when assigning a value of type `xt` to a variable of type `t`.

## Preprocess

Added some checks to prevent the disallowd usage of blank identifiers in `Preprocess` function level. Theses checks are performed at different stages of the preprocessing:

1. `TRANS_ENTER` for `AssignStmt`:
    - Checks if both LHS and RHS are blank identifiers in a `DEFINE` statement.
2. `TRANS_LEAVE` for `NameExpr`:
    - Checks if blank identifier is used as a value in disallowed contexts (excluding `TRANS_ASSIGN_LHS`, `TRANS_RANGE_KEY` and `TRANS_RANGE_VALUE`).
3. `TRANS_LEAVE` for `AssignStmt`:
    - Checks if RHS is a blank identifier when LHS is not, in a `DEFINE` statement.

When any of these conditions are met, the function throws an panics like go message.

<!-- please provide a detailed description of the changes made in this pull request. -->

<details><summary>Contributors' checklist...</summary>

- [X] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [X] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

deelawn reviewed (changes requested):
Thanks Joon.  Looks mostly good as it fixes the example in the issue.  Can you also make a change that panics if the RHS identifier is `_` during a DEFINE assignment?  Maybe assert it is a named expression and compare its name to `blankIdentifier` if it is.
```go
package main

func main() {
	a := _
}
```
And while you are making changes regarding the blank identifier, maybe we want to handle this case as well.  It is technically a valid expression, though I can't think of a use case, so I'm not sure if we should fix it for any reason other than to more strictly adhere to the go spec.
```go
package main

type zilch interface{}

func main() {
	_ = zilch(nil)
}
```
It would be good to get @ltzmaxwell 's review on this as well.

notJoon commented:
@deelawn I've just finished up due to working on other tasks. Now I've modified the code based on your review.

This cases seemed to require handling at the `Preprocess` function level, so i update that function. Additionally, for the last, case, I also couldn't think of a use case ither, but I thought it would be better to follow the go spec. So I fixed that as well.

ltzmaxwell commented:
consider this one:
```go
package main

func main() {
	var a = _
}
```

ltzmaxwell commented on gnovm/pkg/gnolang/preprocess.go:
```diff
@@ -1844,6 +1855,20 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 				n.AssertCompatible(store, last)
 				// NOTE: keep DEFINE and ASSIGN in sync.
 				if n.Op == DEFINE {
+					if n.Op == DEFINE {
```
why?

notJoon replied:
I probably pressed the wrong key 😅

ltzmaxwell commented on gnovm/pkg/gnolang/preprocess.go:
```diff
@@ -430,10 +430,15 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 
 				if n.Op == DEFINE {
-					for _, lx := range n.Lhs {
+					for i, lx := range n.Lhs {
```
Why is it necessary to conduct checks on different sites at various stages? can you provide more tests to support this?

notJoon replied:
After checking again, it wasn't particulary relevant. so removed it.
https://github.com/gnolang/gno/pull/2699/commits/9b8f6971131523a4af78ca647ad8c3d7ee5c1319

ltzmaxwell commented on gnovm/tests/files/typeassert11.gno:
```diff
@@ -0,0 +1,9 @@
+	_ = zilch(nil)
+}
+
+// Output:
```
how about just print something to indicate it's working? like `ok` 😆

notJoon replied:
add some fancy print 
https://github.com/gnolang/gno/pull/2699/commits/9b8f6971131523a4af78ca647ad8c3d7ee5c1319

ltzmaxwell commented:
there should be a better PR title too.

notJoon commented:
@ltzmaxwell I apologize for the delay. I've made some changes to reflect your comments. Still, I can't think of a good idea for the PR title for now. Thanks!

deelawn commented on gnovm/pkg/gnolang/type_check.go:
```diff
@@ -288,6 +288,9 @@ func checkAssignableTo(xt, dt Type, autoNative bool) error {
 	}
 	// case0
 	if xt == nil { // see test/files/types/eql_0f18
+		if dt == nil || dt.Kind() == InterfaceKind {
```
Can you provide a test showing why the second half of this conditional is necessary?

notJoon replied:
Added. please check the `type_check_test.go` file https://github.com/gnolang/gno/pull/2699/commits/e9c055de035d9feada6eeeba4a929685764ed647#diff-ed2a83286fe8c8215a33b8569bb74791d51ae4e0800ef81c326c55f16a85123c

deelawn commented on gnovm/pkg/gnolang/preprocess.go:
```diff
@@ -2892,10 +2898,19 @@ func doConvertType(store Store, last BlockNode, x *Expr, t Type) {
+//
+// This function also checks for the use of blank identifier "_" as a value or type,
+// which is not allowed. If both xt and t are nil, it panics with an appropriate error message.
+func isNamedConversion(xt, t Type, last BlockNode) bool {
```
Was it because you were using the block nodes location before? If so, why did you remove it?

notJoon replied:
this was occurred from wrong auto suggestion. now removed

https://github.com/gnolang/gno/pull/2699/commits/e9c055de035d9feada6eeeba4a929685764ed647

deelawn reviewed (commented):
Looks pretty good. Just a few more questions before approving

zivkovicmilos commented:
Pinging @deelawn and @ltzmaxwell for visibility

ltzmaxwell commented on gnovm/pkg/gnolang/preprocess.go:
```diff
@@ -2892,10 +2898,19 @@ func doConvertType(store Store, last BlockNode, x *Expr, t Type) {
+// This function also checks for the use of blank identifier "_" as a value or type,
+// which is not allowed. If both xt and t are nil, it panics with an appropriate error message.
 func isNamedConversion(xt, t Type) bool {
+	if xt == nil && t == nil {
+		panic("cannot use _ as value or type")
```
did not see any test targeting this.

notJoon replied:
https://github.com/gnolang/gno/pull/2699/commits/d8d53a11c81dab6e6e259df9b0d5489eeabec90c

I wrote new test for `isNamedConversion` function to replace the tests that were previously run through the machine. I checked that this cover everything.

ltzmaxwell replied:
can you provide a file test like this:
```go
package main

type (
	nat  []word
	word int
)

func main() {
	var a nat
	b := []word{0}
	a = b

	println(a)
}
```

to better illustrate why are these logic(added) needed?

ltzmaxwell replied:
what I mean here is that I did not find a situation that xt and t are both nil in the context of `isNamedConversion`.

notJoon replied:
It was definitely an unnecessary check. I remember adding it as defensive logic just in case, but I will delete this. 👍

jefft0 commented:
Removed the "review team" label because this is already reviewed by core devs.

omarsy reviewed (commented):
Maybe you should handle the type case:

```go
package main

func main() {
	var i _
	println(i)
}

```

omarsy commented:
I think your PR will  fix this issue https://github.com/gnolang/gno/issues/1461 too

ltzmaxwell commented on gnovm/pkg/gnolang/preprocess.go:
```diff
@@ -3689,6 +3703,11 @@ func tryPredefine(store Store, last BlockNode, d Decl) (un Name) {
 		})
 		d.Path = last.GetPathForName(store, d.Name)
 	case *ValueDecl:
+		if d.Type != nil {
```
https://github.com/gnolang/gno/blob/ed919917ed1c802448487b8b4a3602db573606aa/gnovm/pkg/gnolang/preprocess.go#L501-L502

do you think these logics can be integrated to avoid excessive fragmentation?

notJoon replied:
sine the conditions being checked are subtly different, so it seems difficult to integrate them into one. but I did modify it to use `isBlankIdentifier` function instead check manually.

https://github.com/gnolang/gno/pull/2699/commits/14543c975b9da00123f2b640655286dd75338361

ltzmaxwell commented:
hey @notJoon, do you have time to take a look at the comments, so we can push this forward? thank you.

ltzmaxwell commented on gnovm/pkg/gnolang/preprocess.go:
```diff
@@ -3075,10 +3081,19 @@ func doConvertType(store Store, last BlockNode, x *Expr, t Type) {
 	}
+	if xt == nil {
+		xt = t
+	}
```
also this.

notJoon replied:
I deleted this part as it seemed unnecessary. However, I added the suggested test as `typeassert10.gno`.

[1] https://github.com/gnolang/gno/pull/2699#discussion_r1845794927
[2] https://github.com/gnolang/gno/pull/2699#discussion_r1822719965

notJoon commented on gnovm/tests/files/typeassert10.gno:
```diff
@@ -0,0 +1,8 @@
+package main
```
how about "blankidentifier"?

ltzmaxwell replied:
ok

notJoon replied:
changed the test names

https://github.com/gnolang/gno/pull/2699/commits/a14935abc1703b216ea3acfc28a4cbaa957f128d

notJoon commented on gnovm/tests/files/typeassert12.gno:
```diff
@@ -0,0 +1,9 @@
+}
+
+// Error:
+// main/files/typeassert12.gno:4:6: cannot use _ as value or type
```
fixed
https://github.com/gnolang/gno/pull/2699/commits/14543c975b9da00123f2b640655286dd75338361

mvertes reviewed (approved):
Ok to me

ltzmaxwell commented on gnovm/pkg/gnolang/preprocess.go:
```diff
@@ -3607,10 +3616,12 @@ func doConvertType(store Store, last BlockNode, x *Expr, t Type) {
+	// or when types are not fully resolved during preprocessing.
+	// relative test: gnovm/tests/files/blankidentifier7.gno
+	if xt == nil || t == nil {
+		return false
```
please revert this as no sufficient reason for the modification. @notJoon

ltzmaxwell replied:
It would be good to merge after this.

notJoon replied:
In the current modification, a panic is occurred in `blankidentifier7` test if I remove this nil check. The panic occurs in `isNamedConversion()` due to a nil pointer derefernce when attempting to call `IsNamed()` on a nil type.

I think this happens during type conversion preprocessing where either `x` and `xt` parameters are nil.

In particular, if `x` or `xt` is nil, this condition tries to call `nil.IsNamed()`, which causes the nil pointer problem.
https://github.com/gnolang/gno/blob/2093d8a43c1ec632707f28021860ac52f5a80c26/gnovm/pkg/gnolang/preprocess.go#L3638-L3646

So the possible solutions are:

1. add `nil` check condition (current version):

```go
func isNamedConversion(xt, t Type) bool {
    if xt == nil || t == nil { // <-
        return false
    }
    
    // ...

    if !c1 && !c2 {
        if t.IsNamed() && !xt.IsNamed() ||
            !t.IsNamed() && xt.IsNamed() {
            return true
        }
    }
    return false
}
```

2. Check `nil` on calling side:

```go
func convertType(store Store, last BlockNode, x *Expr, t Type) {
    if t == nil {
        return // <-
    }
    // ...
}
```

I'm not sure why `x` or `xt` becomes `nil` when the parameter is named blank identifier, but here's my guess.

```go
func Foo(_ string) string { /* ... */ }
```

1. The blank identifier typically indicates to the compiler that a parameter won't be used within the function.
2. However, this becomes problematic because:
    - `isNamedConversion()` is called during the preprocess stage
    - At this stage, the compiler is processing type information
    - It appears that type information for parameters with this identifier is either not properly transmitted or omitted

These are the causes and solutions I came up with. Please let me know if theres other solutions or if I'm wrong. thank you!

ltzmaxwell replied:
not necessary after discussion with @notJoon
